src/kafka/connection/connection_manager.hh

```diff
@@ -44,15 +44,62 @@ private:
     std::unordered_map<connection_id, lw_shared_ptr<kafka_connection>, boost::hash<connection_id>> _connections;
     std::string _client_id;
 
+    semaphore _connect_semaphore;
+    semaphore _send_semaphore;
+
 public:
 
     explicit connection_manager(std::string client_id)
-        : _client_id(std::move(client_id)) {};
+        : _client_id(std::move(client_id)),
+        _connect_semaphore(1),
+        _send_semaphore(1) {}
 
     future<lw_shared_ptr<kafka_connection>> connect(const std::string& host, uint16_t port);
     std::optional<lw_shared_ptr<kafka_connection>> get_connection(const connection_id& connection);
     future<> disconnect(const connection_id& connection);
 
+    template<typename RequestType>
+    future<typename RequestType::response_type> send(RequestType request, const std::string& host, uint16_t port) {
+        // In order to preserve ordering of sends, a semaphore with
+        // count = 1 is used due to its FIFO guarantees.
+        //
+        // It is important that connect() and send() are done
+        // with semaphore, as naive implementation
+        //
+        // connect(host, port).then([](auto conn) { conn.send(req1); });
+        // connect(host, port).then([](auto conn) { conn.send(req2); });
+        //
+        // could introduce reordering of requests: after both
+        // connects resolve as ready futures, the continuations (sends)
+        // are not guaranteed to run with any specific order.
+        //
+        // In order to not limit concurrency, send_future is
+        // returned as future<future<response>> and "unpacked"
+        // outside the semaphore - scheduling inside semaphore
+        // (only 1 at the time) and waiting for result outside it.
+        return with_semaphore(_send_semaphore, 1, [this, request = std::move(request), host, port] {
+            return connect(host, port).then([request = std::move(request)](auto conn) {
```

Comment:
Why do we connect here again? Shouldn't this part be taken care of beforehand with `connect` method? If not, then is the aforementioned method still useful?

Reply:
Another question would be: don't we want to know that sending failed due to broken connection? Here, while connecting before each send we loose that info.

Reply:
> Why do we connect here again? Shouldn't this part be taken care of beforehand with connect method? If not, then is the aforementioned method still useful?

1. Ordering:
```
connection_manager.connect(HOST1).then([](auto conn){
    conn->send(MSG1);
});
connection_manager.connect(HOST1).then([](auto conn){
    conn->send(MSG2);
});
```
`MSG1` and `MSG2` will not be sent in correct order, as both `connect` return instantly (situation when connection to `HOST1` established beforehand) and two `then` can run in any order. (That is why I use semaphores).

2. Ease of use, high-level API: In most usages of `connection_manager`, we want to easily send a message to some broker, not "manually" manage connections.

> don't we want to know that sending failed due to broken connection

A general `NETWORK_EXCEPTION` error code is set in this situation: https://github.com/avelanarius/seastar/pull/14/files/ada9c5e22c0dc32242018b3489a3de57470d1539#diff-d7bfcd1b2f043e7a26b70b26a7098586R97

We should also `disconnect` the broken connection in the `send` method (not implemented currently).

Reply:
Makes sense. Just add a TODO in the code if you're not implementing that in this PR